sdk/storage/azure-storage-blob/azure/storage/blob/_download.py

```diff
@@ -531,81 +529,112 @@ def chunks(self):
-            if self._encryption_options.get("key") is not None or self._encryption_options.get("resolver") is not None:
-                data_start = (self._start_range or 0) + len(self._current_content)
+            end = self._download_start + self._total_size
 
             iter_downloader = _ChunkDownloader(
                 client=self._clients.blob,
                 non_empty_ranges=self._non_empty_ranges,
                 total_size=self.size,
                 chunk_size=self._config.max_chunk_get_size,
-                current_progress=self._first_get_size,
-                start_range=data_start,
-                end_range=data_end,
-                stream=None,
-                parallel=False,
+                current_progress=current_progress,
+                start_range=start,
+                end_range=end,
                 validate_content=self._validate_content,
                 encryption_options=self._encryption_options,
                 encryption_data=self._encryption_data,
                 use_location=self._location_mode,
                 **self._request_options
             )
+
+        initial_content = self._current_content if self._first_chunk else b''
         return _ChunkIterator(
-            size=self.size,
-            content=self._current_content,
+            size=self._total_size,
+            content=initial_content,
             downloader=iter_downloader,
-            chunk_size=self._config.max_chunk_get_size)
+            chunk_size=self._chunk_size)
 
-    def read(self, size: Optional[int] = -1) -> T:
+    def read(self, size: int = -1, *, chars: Optional[int] = None) -> T:
         """
-        Read up to size bytes from the stream and return them. If size
-        is unspecified or is -1, all bytes will be read.
+        Read the specified bytes or chars from the stream. If `encoding`
+        was specified on `download_blob`, it is recommended to use the
+        chars parameter to read a specific number of chars to avoid decoding
+        errors. If size/chars is unspecified or negative all bytes will be read.
 
-        :param Optional[int] size:
+        :param int size:
             The number of bytes to download from the stream. Leave unspecified
-            or set to -1 to download all bytes.
+            or set negative to download all bytes.
+        :param Optional[int] chars:
+            The number of chars to download from the stream. Leave unspecified
+            or set negative to download all chars. Note, this can only be used
+            when encoding is specified on `download_blob`.
         :returns:
             The requested data as bytes or a string if encoding was specified. If
             the return value is empty, there is no more data to read.
         :rtype: T
         """
-        if size == -1:
-            return self.readall()
         # Empty blob or already read to the end
         if size == 0 or self._offset >= self.size:
```

Comment:
Yep, done. I also needed to change the "already read to end" condition to not use `offset` (now renamed to `read_offset` because that value can be in bytes or chars and `self.size` is in bytes only.